modem/tones.c

```diff
@@ -52,23 +57,66 @@ struct _ModemTonesPrivate
   ModemTonesStoppedNotify *notify;
   gpointer data;
 
-  GQueue stop_requests[1];
+  GHashTable *notify_map;
 
   unsigned dispose_has_run:2;
 };
 
+struct _ModemNotifyData
+{
+  uint32_t event_id;
+  ModemTonesStoppedNotify *notify;
+  ModemTones *modem;
+  gpointer data;
+};
+
+typedef struct _ModemNotifyData ModemNotifyData;
+
+static void
+ngf_callback(NgfClient *client, uint32_t id, NgfEventState state, void *userdata)
+{
+  ModemTones *self = (ModemTones*) userdata;
+  ModemNotifyData *notify_data = NULL;
+
+  switch (state) {
+    /* Both FAILED and COMPLETED mean that the event isn't playing anymore.
+     * So with both cases check if tone user wants to be notified. */
+    case NGF_EVENT_FAILED:
+      DEBUG("NGFD event failed.");
+      /* fall through */
+    case NGF_EVENT_COMPLETED:
+      notify_data = g_hash_table_lookup(self->priv->notify_map, &id);
+      break;
+
+    case NGF_EVENT_PLAYING:
+      break;
+
+    case NGF_EVENT_PAUSED:
+      break;
+
+  }
+
+  if (notify_data) {
+    notify_data->notify(self, notify_data->event_id, notify_data->data);
+    g_hash_table_remove(self->priv->notify_map, &id);
+  }
+}
+
 static void
 modem_tones_init(ModemTones *self)
 {
   self->priv = G_TYPE_INSTANCE_GET_PRIVATE(
     self, MODEM_TYPE_TONES, ModemTonesPrivate);
+  self->priv->connection = dbus_g_bus_get(DBUS_BUS_SYSTEM, NULL);
   self->priv->timer = g_timer_new();
-  self->priv->proxy =
-    dbus_g_proxy_new_for_name(dbus_g_bus_get(DBUS_BUS_SYSTEM, NULL),
-      "com.Nokia.Telephony.Tones",
```

Comment:
Are these strings defined in some header files that we can use instead of directly here?